src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -414,6 +431,9 @@ private void AddLexedTokenSlot()
                 if (shiftCount > 0)
                 {
                     Array.Copy(_lexedTokens, shiftOffset, _lexedTokens, 0, shiftCount);
+
+                    // Clear out the cached tokens from shiftCount to the end of the token array
+                    Array.Clear(_lexedTokens, shiftCount, shiftOffset);
```

Comment:
I don't think I understand why we need to do this. I don't think it's _wrong_, per se, but I'm not following what in the new logic would require this clear now when it wasn't before.

Reply:
It needs to be done now because the Dispose method only does the Array.Clear from 0 -> _tokenCount, not the whole array before releasing the array back into the pool.

Reply:
> because the Dispose method only does the Array.Clear from 0 -> _tokenCount

What's the reasoning behind not clearing the whole array? In theory, `Array.Clear` should just turn into `memset`, which I wouldn't expect to take very long, even on a 4KB array.

Reply:
The benchmark for the small file case showed a significant regression  if I did the Array.Clear over the whole array.

Reply:
Gotcha, thanks for the clarification.

Reply:
The manipulations done around _tokenCount and _tokenOffset are complicated enough that I don't feel completely comfortable with assuming that _tokenCount is the proper length to clear out. (I hadn't seen it's manipulation in Reinitialize and how it can be used to track into _blendedTokens too). Instead, I think it will be clearer and safer to just keep track of the largest index written into the lexedTokenArray, and just clear that amount out on dispose.